pbs_light.go

```diff
@@ -462,6 +413,63 @@ func auction(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
 	mRequestTimer.UpdateSince(pbs_req.Start)
 }
 
+func sortBidsAddKeywordsMobile(bids pbs.PBSBidSlice, pbs_req *pbs.PBSRequest, priceGranularitySetting string) {
+	if priceGranularitySetting == "" {
+		priceGranularitySetting = DEFAULT_PRICE_GRANULARITY
+	}
+
+	// record bids by ad unit code for sorting
+	code_bids := make(map[string]pbs.PBSBidSlice, len(bids))
+	for _, bid := range bids {
+		code_bids[bid.AdUnitCode] = append(code_bids[bid.AdUnitCode], bid)
+	}
+
+	// loop through ad units to find top bid
+	for _, unit := range pbs_req.AdUnits {
+		bar := code_bids[unit.Code]
+
+		if len(bar) == 0 {
+			if glog.V(1) {
+				glog.Infof("No bids for ad unit '%s'", unit.Code)
+			}
+			continue
+		}
+		sort.Sort(bar)
+
+		// after sorting we need to add the ad targeting keywords
+		for i, bid := range bar {
+			priceBucketStringMap := pbs.GetPriceBucketString(bid.Price)
+			roundedCpm := priceBucketStringMap[priceGranularitySetting]
+
+			hbPbBidderKey := "hb_pb_" + bid.BidderCode
+			hbBidderBidderKey := "hb_bidder_" + bid.BidderCode
+			hbCacheIdBidderKey := "hb_cache_id_" + bid.BidderCode
+			if pbs_req.MaxKeyLength != 0 {
+				hbPbBidderKey = hbPbBidderKey[:min(len(hbPbBidderKey), int(pbs_req.MaxKeyLength))]
+				hbBidderBidderKey = hbBidderBidderKey[:min(len(hbBidderBidderKey), int(pbs_req.MaxKeyLength))]
+				hbCacheIdBidderKey = hbCacheIdBidderKey[:min(len(hbCacheIdBidderKey), int(pbs_req.MaxKeyLength))]
+			}
+			pbs_kvs := map[string]string{
+				hbPbBidderKey:      roundedCpm,
+				hbBidderBidderKey:  bid.BidderCode,
+				hbCacheIdBidderKey: bid.CacheID,
+			}
+			// For the top bid, we want to add the following additional keys
+			if i == 0 {
+				pbs_kvs["hb_pb"] = roundedCpm
+				pbs_kvs["hb_bidder"] = bid.BidderCode
+				pbs_kvs["hb_cache_id"] = bid.CacheID
+				if bid.BidderCode == "audienceNetwork" {
+					pbs_kvs["hb_creative_load_method"] = "demand_sdk"
+				} else {
+					pbs_kvs["hb_creative_load_method"] = "html"
```

Comment:
Can we turn these magic strings into constants?

All that stuff you explained in slack would (if you can make it a bit more concise) be great to have in go-docs.